frame/contracts/src/storage.rs

```diff
@@ -250,48 +252,35 @@ impl<T: Config> ContractInfo<T> {
 			return weight_limit
 		}
 
-		let mut queue = <DeletionQueue<T>>::get();
+		while remaining_key_budget > 0 {
+			let Some(entry) = queue.next() else { break };
 
-		while !queue.is_empty() && remaining_key_budget > 0 {
-			// Cannot panic due to loop condition
-			let trie = &mut queue[0];
 			#[allow(deprecated)]
 			let outcome = child::kill_storage(
-				&ChildInfo::new_default(&trie.trie_id),
+				&ChildInfo::new_default(&entry.contract().trie_id),
 				Some(remaining_key_budget),
 			);
-			let keys_removed = match outcome {
+
+			match outcome {
 				// This happens when our budget wasn't large enough to remove all keys.
-				KillStorageResult::SomeRemaining(c) => c,
-				KillStorageResult::AllRemoved(c) => {
-					// We do not care to preserve order. The contract is deleted already and
-					// no one waits for the trie to be deleted.
-					queue.swap_remove(0);
-					c
+				KillStorageResult::SomeRemaining(keys_removed) => {
+					remaining_key_budget = remaining_key_budget.saturating_sub(keys_removed);
+					break
```

Comment:
I assume the loop was breaking as well before at the end of this this iteration, when there was too little budget to kill the entire contract storage, feels a bit safer to break here directly

Reply:
What I understand from the previous implementation is that the loop would try to kill some storage from following tries (now entries) in the queue with the remaining budget even if this budget wasn't enough to kill all the storage from the current trie (entry).
By breaking the loop we are not trying to kill more storage from further elements in the queue.

Reply:
this is still doing that kill_storage will try to kill up to `remaining_budget` keys from storage.
if we did not have enough budget to remove it all, we break from the loop, at that point I assume that remaining_key_budget is 0

we can simplify it to that maybe to make the code more explicit

```rust
			match outcome {
				// This happens when our budget wasn't large enough to remove all keys.
				KillStorageResult::SomeRemaining(_) => return weight_limit,
				KillStorageResult::AllRemoved(keys_removed) => {
					entry.remove();
					remaining_key_budget = remaining_key_budget.saturating_sub(keys_removed);
				},
			};
```